source/Pages/jupyter_notebooks.rst

```diff
@@ -7,6 +8,147 @@ Jupyter Notebooks
+
+=====================
+Singularity container
+=====================
+
+Pre-built container
+===================
+
+To run a notebook in a singularity container, we have to fetch or build the container first. A tutorial on containers can be found in :ref:`singularity-building`, but note that this focuses on using GPUs. A more general introduction is provided here.
+
+First we start by fetching a container:
+
+.. code-block:: bash
+
+   singularity build jupyter.sif docker://jupyter/scipy-notebook:latest
+
+This will pull one of the official jupyter containers from docker hub, and build a singularity container from it. This container encapsulates the entire environment and can be entered to start a notebook session. More official containers can be found `here <https://hub.docker.com/r/jupyter/>`_, and more docker images in general can be found at `docker hub <https://hub.docker.com/>`_.
+
+After the build procedure is complete, you can start the jupyter instance with 
+
+.. code-block:: bash
+
+   singularity run jupyter.sif
+
+which will automatically start the instance. Alternatively, you can start an interactive shell session in the container and start it manually:
+
+.. code-block:: bash
+
+   singularity shell jupyter.sif
+   jupyter lab
+
+To receive the notebook locally in your browser, as mentioned above, a tunnel has to be opened in a new terminal, with:
+
+.. code-block:: bash
+   
+   ssh -NL 8888:wn-db-01:8888 spider
+
+Where, again, the machine name and port name have to match where you are running the job and the port chosen by jupyter, respectively. Now you can open the link provided by jupyter, which has the shape of ``http://localhost:8888/lab?token=abc123``.
+
+If the forwarding or other steps do not work, please contact :ref:`our helpdesk <helpdesk>`.
+
+Custom image
+============
+
+Images can be customised to suit your needs, by adding extra steps during the build process. This is done with so-called 'definitions' files. These are plaintext files with instructions for the singularity build. For a full overview, see the `singularity documentation <https://docs.sylabs.io/guides/latest/user-guide/definition_files.html>`_. Here is a small example that can be expanded. Make a file called ``jup-custom.def`` and fill it with:
+
+.. code-block:: bash
+
+   Bootstrap: docker
+   From: jupyter/scipy-notebook:latest
+
+   %post
+     pip install docopt
+
+   %runscript
+     jupyter lab --ip=0.0.0.0
+
+   %help
+     This is a demo container to show how to run jupyter lab 
+
+You can build this with:
+
+.. code-block:: bash
+
+   singularity build jup-custom.sif jup-custom.def
+
+and once it is finished building, you can enter the `sif` file with the ``singularity shell`` command, or start jupyter directly with ``singularity run``. You still have to forward the connection as described above before you can open the notebook in a browser.
```

Comment:
Just wondering if its clear to a reader what of the above def file is custom, May perhaps worth to spend one sentence on pointing that out explicitly.